docker-compose.yml

```diff
@@ -1221,6 +1222,50 @@ services:
     volumes: *conda-volumes
     command: *python-conda-command
 
+  conda-python-no-numpy:
+    # Usage:
+    #   docker-compose build conda
+    #   docker-compose build conda-cpp
+    #   docker-compose build conda-python
+    #   docker-compose build conda-python-no-numpy
+    #   docker-compose run --rm conda-python-no-numpy
+    image: ${REPO}:${ARCH}-conda-python-${PYTHON}-no-numpy
+    build:
+      context: .
+      dockerfile: ci/docker/conda-python-pandas.dockerfile
+      cache_from:
+        - ${REPO}:${ARCH}-conda-python-${PYTHON}-pandas-${PANDAS}
+      args:
+        repo: ${REPO}
+        arch: ${ARCH}
+        python: ${PYTHON}
+        numpy: ${NUMPY}
+        pandas: ${PANDAS}
+    shm_size: *shm-size
+    environment:
+      <<: [*common, *ccache, *sccache]
+      PARQUET_REQUIRE_ENCRYPTION:  # inherit
+      # The without numpy mark is only used to select tests to be run
+      # when numpy is not installed but those tests will also run
+      # if numpy is present.
```

Comment:
But they would run anyway if they are not marked with `pytest.mark.numpy`, right? What is the point of this specific option?

Reply:
I understand the concerns but I think the alternative solution that's been chosen (run a small subset of the tests) is much worse. Why wouldn't we exercise reading CSV and JSON files, for example?

I also find concerning that this is hidden somewhere in the `docker-compose.yml` (many developers will never go look there) instead of declared in plain sight at the top of the corresponding test files.

For example, if you don't want to run `test_csv.py` when NumPy is absent (assuming there's a good reason for it), you could add an explicit skip at the top of that file. The maintenance burden would be small enough.

Reply:
> Why wouldn't we exercise reading CSV

I know it was only an example, but FWIW it would be very easy to enable the CSV tests in this PR. The `test_csv.py` file does import numpy, but it only uses it once to generate random integers. That's something we should be able to replace with stdlib `random` quite easily I think.

Reply:
Still, it would be better if the skips were done at the top of the test files themselves, instead of passing an environment variable from `docker-compose.yml`.

Reply:
I don't know, if I add `test_csv` manually to `PYTEST_PYARGS`, things seem to work for me (though no CSV test gets run because of missing `without_numpy` markers):
```
+ pytest -r s -k without_numpy --pyargs pyarrow.tests.test_csv pyarrow.tests.test_compute pyarrow.tests.test_dataset pyarrow.tests.test_flight pyarrow.tests.test_without_numpy

[...]

==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 981 items / 882 deselected / 99 selected                                                                                                          

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_compute.py ..............                                                         [ 14%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_dataset.py .........                                                              [ 23%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_flight.py .....s.............s.....s...........s.......ss..................ss.s.. [ 94%]
..                                                                                                                                                    [ 96%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_without_numpy.py sss                                                              [100%]

================================================================== short test summary info ==================================================================
SKIPPED [9] opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/conftest.py:94: slow NOT enabled
SKIPPED [3] opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/conftest.py:94: nonumpy NOT enabled
================================================ 87 passed, 12 skipped, 882 deselected, 13 warnings in 7.09s ================================================
```

Reply:
And even if I clear `PYTEST_PYARGS`, the tests still pass:
```
+ pytest -r s -k without_numpy --pyargs pyarrow
/opt/conda/envs/arrow/lib/python3.8/site-packages/pytest_cython/__init__.py:2: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources import get_distribution
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:2529: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(parent)
/opt/conda/envs/arrow/lib/python3.8/site-packages/pkg_resources/__init__.py:3117: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 7634 items / 7527 deselected / 3 skipped / 107 selected                                                                                           

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_array.py ........                                                                 [  7%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_compute.py ..............                                                         [ 20%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_dataset.py .........                                                              [ 28%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_flight.py .....s.............s.....s...........s.......ss..................ss.s.. [ 95%]
..                                                                                                                                                    [ 97%]
opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_without_numpy.py sss                                                              [100%]

```

Reply:
Ah, but I forgot to uninstall NumPy...

Reply:
Ok, now I have:
```diff
diff --git a/python/pyarrow/tests/test_csv.py b/python/pyarrow/tests/test_csv.py
index 9ddb5197e9..3a77f1a2e2 100644
--- a/python/pyarrow/tests/test_csv.py
+++ b/python/pyarrow/tests/test_csv.py
@@ -36,7 +36,12 @@ import weakref
 
 import pytest
 
-import numpy as np
+pytestmark = pytest.mark.numpy
+
+try:
+    import numpy as np
+except ImportError:
+    np = None
 
 import pyarrow as pa
 from pyarrow.csv import (
```

And then I can collect `test_csv` without NumPy:
```
+ pytest -r s -k without_numpy --pyargs pyarrow.tests.test_array pyarrow.tests.test_csv

[...]

==================================================================== test session starts ====================================================================
platform linux -- Python 3.8.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /
plugins: hypothesis-6.104.1, cython-0.2.2
collected 420 items / 412 deselected / 8 selected                                                                                                           

opt/conda/envs/arrow/lib/python3.8/site-packages/pyarrow/tests/test_array.py ........                                                                 [100%]

````

Reply:
By the way, the fact that only those tests listed in `PYTEST_PYARGS` *and* with a `pytest.mark.without_numpy` marker are executed in this instance is clearly neither maintainable nor future-proof (we'll easily forget about adding these when adding new tests).

Reply:
> And then I can collect test_csv without NumPy:

`test_csv.py` is an easy example because it hardly uses numpy in practice (as I mentioned above, there is exactly one use of it to generate random integers, that can be easily be fixed in this PR). But the problem is for files that use numpy in `pytest.mark.parametrize`, as far as I understand

Reply:
yeah, `test_convert_builtin.py` is a challenge for example

Reply:
I've just tested this incantation and it seems to workaround the test collection problem (unlike `pytestmark = ...` which would still execute the module to the end):
```python
try:
    import numpy as np
except ImportError:
    pytest.skip(reason="blah", allow_module_level=True)
```

Reply:
> I've just tested this incantation and it seems to workaround the test collection problem

Ah, that would be a nice solution!

Reply:
this is indeed really nice! Thanks @pitrou I'll work with that! Will ping you once I think this is again ready for review!
Thanks both for the constructive feedback and help!

Reply:
Thanks for your work @raulcd :-)